### ISSUE TITLE
Dockerfile is missing HEALTHCHECK

### ISSUE THREAD

**Issue by kare** [ipfs/go-ipfs#8404]
### Checklist

- [X] My issue is specific & actionable.
- [X] I am not suggesting a protocol enhancement.
- [X] I have searched on the [issue tracker](https://github.com/ipfs/go-ipfs/issues?q=is%3Aissue) for my issue.

### Description

`Dockerfile` is missing `HEALTHCHECK`. It would be a nice addition. Something simple should work like checking whether exposed ports respond with curl or netcat. https://docs.docker.com/engine/reference/builder/#healthcheck

**Comment by arjunraghurama** [comment]
Hello,
Which endpoint to query in order check if the container is healthy or not?
I hope it is localhost:8080

**Comment by kare** [comment]
Container exposes several ports both tcp and udp. See [Dockerfile](https://github.com/ipfs/go-ipfs/blob/65d570c6cbe24f313fc6f02c3dc16aabc53306d1/Dockerfile#L75-L84) for details.

```
# Swarm TCP; should be exposed to the public
EXPOSE 4001
# Swarm UDP; should be exposed to the public
EXPOSE 4001/udp
# Daemon API; must not be exposed publicly but to client services under you control
EXPOSE 5001
# Web Gateway; can be exposed publicly with a proxy, e.g. as https://ipfs.example.org
EXPOSE 8080
# Swarm Websockets; must be exposed publicly when the node is listening using the websocket transport (/ipX/.../tcp/8081/ws).
EXPOSE 8081
```

Checking ports 4001/tcp, 4001/udp, and 5001/tcp would be a good start.
Secondary would be a Websocket connection to port 8081/tcp and HTTP connection to port 8080/tcp.

Am I right in thinking @Stebalien?

**Comment by Stebalien** [comment]
I would check if some known file can be retrieved from the gateway. E.g., the empty directory. That should provide a reasonable end-to-end health check.

**Comment by arjunraghurama** [comment]
What about http://localhost:5001/webui ?

**Comment by Stebalien** [comment]
That will stall until you actually download it. The best bets are either an "empty directory" (which you're guaranteed to have) or the welcome docs (see the output of `ipfs init`).

**Comment by arjunraghurama** [comment]
Hey @Stebalien  I found below file is guaranteed to exist after a node has started.
`QmYwAPJzv5CZsnA625s3Xf2nemtYgPpHdWEz79ojWnPbdG/readme`. Is this fine??

**Comment by Stebalien** [comment]
As long as that CID isn't hard-coded because it _can_ change and a hard-coded CID will bit rot.

**Comment by arjunraghurama** [comment]
Yeah. The contents of `readme` file might change in future resulting in change of CID. I couldn't figure out the path of readme file inside the container.
`ipfs id` command will return node info once it is up. Can this be used?

**Comment by arjunraghurama** [comment]
Or we could ping any one of the node in the swarm?

**Comment by Stebalien** [comment]
> `ipfs id` command will return node info once it is up. Can this be used?

That would work, but it won't verify that data can be fetched from the repo.

> Or we could ping any one of the node in the swarm?

That's likely going to be flaky.

**Comment by arjunraghurama** [comment]
I tried this within the container 
`hash=`echo "I <3 IPFS -$(whoami)" | ipfs add -q`
wget "https://ipfs.io/ipfs/$hash" -O /dev/null &> /dev/null` but wget sometimes fails fetch the file

**Comment by Stebalien** [comment]
I'd be careful about _adding_ things as that can trigger network traffic, etc.

In terms of that failure, you're trying to fetch them from the gateway. Try to keep the health check locally.

Again, just fetch an empty directory. It'll always have the same CID and every node will have it.

**Comment by arjunraghurama** [comment]
Well in that case, we would need to create an empty directory that no one will add any data into and add directory to ipfs and then fetch locally. Is this flow correct?

**Comment by Stebalien** [comment]
The CID of the empty directory is QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn. It's pinned on all nodes by default.

**Comment by arjunraghurama** [comment]
Oh. Okay. I was unaware of that. Thanks for the info.

**Comment by amias-channer** [comment]
sorry to dig up an old thread but there is an issue with doing the Docker Healthcheck this way. 

If you configure the API.Authorizations then the api will not be available when this is called because the healthcheck wont have the api credentials or a sensible way to obtain them for the call.  

Is it possible to do this using a specific command that is not restricted by the Api Auth settings ? 

it could use this method underneath the endpoint but just not require auth,  it might need rate limiting or possibly to only available in the first few minutes of run time.
